File: app/demo/src/main/java/net/gotev/recycleradapterdemo/SyncItem.java
package net.gotev.recycleradapterdemo;

import android.content.Context;
import android.support.annotation.NonNull;

import net.gotev.recycleradapter.AdapterItem;

/**
 * @author Aleksandar Gotev
 */

public class SyncItem extends ExampleItem {

    private int id;
    private String suffix;

    public SyncItem(Context context, int id, String suffix) {
        super(context, "item " + id + " " + suffix);
        this.id = id;
        this.suffix = suffix;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }

        if (obj == null || getClass() != obj.getClass()) {
            return false;
        }

        SyncItem other = (SyncItem) obj;
        return id == other.id;
    }

    @Override
    public int compareTo(@NonNull AdapterItem otherItem) {
        if (otherItem.getClass() != getClass())
            return -1;

        SyncItem item = (SyncItem) otherItem;

        if (id == item.id)
            return 0;

        return id > item.id ? 1 : -1;
    }

    @Override
    public boolean hasToBeReplacedBy(AdapterItem newItem) {
        SyncItem otherItem = (SyncItem) newItem;
        return (otherItem.id + suffix.hashCode()) != (id + suffix.hashCode());
    }
}

File: app/demo/src/main/java/net/gotev/recycleradapterdemo/SyncItem.kt
package net.gotev.recycleradapterdemo

import android.content.Context

import net.gotev.recycleradapter.AdapterItem

class SyncItem(context: Context, private val id: Int, private val suffix: String) : ExampleItem(context, "item $id $suffix") {

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        if (other == null || javaClass != other.javaClass) {
            return false
        }

        return id == (other as SyncItem).id
    }

    override fun compareTo(other: AdapterItem<*>): Int {
        if (other.javaClass != javaClass)
            return -1

        val item = other as SyncItem

        if (id == item.id)
            return 0

        return if (id > item.id) 1 else -1
    }

    override fun hasToBeReplacedBy(newItem: AdapterItem<*>): Boolean {
        val otherItem = newItem as SyncItem
        return otherItem.id + suffix.hashCode() != id + suffix.hashCode()
    }
}
